soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Select.scala

```diff
@@ -22,33 +19,109 @@ case class SubSelect(selects: BinaryTree[Select], alias: String)
+}
+case class JoinQuery(selects: BinaryTree[Select], definiteAlias: String) extends JoinSelect {
+  val alias = Some(definiteAlias)
+  override def toString = "(" + Select.toString(selects) + ") AS " + TableName.removeValidPrefix(definiteAlias)
+  def replaceHoles(f: Hole => Expression): JoinQuery =
+    copy(Select.walkTreeReplacingHoles(selects, f))
+  def rewriteJoinFuncs(f: Map[TableName, UDF], aliasProvider: AliasProvider): JoinQuery =
+    copy(selects = Select.rewriteJoinFuncs(selects, f)(aliasProvider))
+  def directlyReferencedJoinFuncs = Select.findDirectlyReferencedJoinFuncs(selects)
+}
+case class JoinFunc(tableName: TableName, params: Seq[Expression])(val position: Position) extends JoinSelect {
+  val alias = tableName.alias
+  override def toString =
+    TableName.withSoqlPrefix(tableName.name) + "(" + params.mkString(", ") + ")" + tableName.aliasWithoutPrefix.map(" AS " + _)
+
+  def directlyReferencedJoinFuncs = Set(tableName.copy(alias = None))
 
-  def selects: Option[BinaryTree[Select]] = {
-    subSelect match {
-      case Left(_) => None
-      case Right(SubSelect(s, _)) => Some(s)
+  def replaceHoles(f: Hole => Expression): JoinFunc =
+    copy(params = params.map(_.replaceHoles(f)))(position)
+
+  def rewriteJoinFuncs(f: Map[TableName, UDF], aliasProvider: AliasProvider): JoinQuery = {
+    val aliasless = tableName.copy(alias=None)
+    f.get(aliasless) match {
+      case Some(udf) =>
+        rewrite(udf, params, aliasProvider).rewriteJoinFuncs(f, aliasProvider)
+      case None =>
+        throw UnknownUDF(aliasless)(position)
     }
   }
 
-  override def toString: String = {
-    val (subSelectStr, aliasStrOpt) = subSelect match {
-      case Right(SubSelect(select, subAlias)) =>
-        val selectStr = Select.toString(select)
-        (s"($selectStr)", Some(subAlias))
-      case Left(tn@TableName(name, alias)) =>
-        (tn.toString, None)
+  def rewrite(udf: UDF, params: Seq[Expression], aliasProvider: AliasProvider): JoinQuery = {
+    // ok we're going to rewrite
+    //   join @foo(x, y, z) as bleh
+    // to
+    //   join (select @expr.:*, @expr.* from (select x,y,z from @single_row) as vars join lateral (body) as expr on true) as bleh;
+    // ..or that's the ideal, anyway.  Turns out you can't FROM a
+    // query in soql, so we'll actually rewrite to "..from @single_row
+    // join (select x,y,z) join lateral (body)" instead.
+    //
+    // Probably will try to turn `from single_row join bleh on true`
+    // into `from bleh` over in the postgres adapter later.
+    if(udf.arguments.length != params.length) {
+      throw MismatchedParameterCount(expected = udf.arguments.length, got = params.length)(position)
     }
+    val expr = aliasProvider("_expr")
+    val vars = aliasProvider("_vars")
+    val singleRowFrom = Some(TableName(TableName.SingleRow))
+    val labelledJoin =
+      Select(
+        distinct = false,
+        selection = Selection(Some(StarSelection(Some(expr), Nil)),
```

Comment:
Does this kind of select :*, * already work in this nested setting in many processes downstream all the way to sqlize?  Would it be simpler to expand * here if downstream processes don't handle it?

Reply:
It seems to, but I wouldn't be super-averse to saying "join functions can't select system cols"